### PR TITLE
rewriting the operation when removing and adding the same partition col with the same transformation 

### PR DESCRIPTION
**Pull request by CodingCat**
this PR is a followup of https://github.com/apache/iceberg/pull/3632#discussion_r776515101 . It rewrites the operation when deleting and adding the same partition col so that users will not encounter an error

specifically:

* when the user removes and adds the partition col with the same transformation and the same name, we ensure it is a no-op
* when the user remove and adds the partition col with the same transformation but DIFFERENT name, we just rewrite the remove and add as rename

We have this scenario that we need to change the partition column of a table with an API call, as said in the last PR, with the functionality in PR, we can just remove all "conflicted" column names without the burden to check whether the transformation is the same

### CODE REVIEW

**Comment by CodingCat** [comment]
Hi, @rdblue would you mind taking a look? thank you!

File: core/src/test/java/org/apache/iceberg/TestUpdatePartitionSpec.java [hunk 1]
```diff
@@ -477,12 +477,12 @@ public void testAddRedundantTimePartition() {
...
-  public void testAddDeletedField() {
-    AssertHelpers.assertThrows("Should fail adding a duplicate field",
-        IllegalArgumentException.class, "Cannot add duplicate partition field",
-        () -> new BaseUpdatePartitionSpec(formatVersion, PARTITIONED)
-            .removeField("shard")
-            .addField(bucket("id", 16))); // duplicates shard
+  public void testNoEffectAddDeletedSameField() {
+    PartitionSpec updated = new BaseUpdatePartitionSpec(formatVersion, PARTITIONED)
+        .removeField("shard")
+        .addField(bucket("id", 16))
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestUpdatePartitionSpec.java:483]: This isn't correct because "shard" is a custom name. The result of separating this into two updates (remove and then add) would be a different partition spec, with the bucket field named `id_bucket_16`.

In this case, I think you just need to detect the rename and rewrite the remove/add to rename.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java [hunk 2]
```diff
@@ -133,6 +133,11 @@ public BaseUpdatePartitionSpec addField(String name, Term term) {
     Pair<Integer, String> validationKey = Pair.of(sourceTransform.first(), sourceTransform.second().toString());
 
     PartitionField existing = transformToField.get(validationKey);
+    if (existing != null && deletes.contains(existing.fieldId()) &&
+        existing.transform().toString().equals(sourceTransform.second().toString())) {
+      deletes.remove(existing.fieldId());
+      return this;
+    }
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:140]: Nit: missing empty line between a control flow block and the following statement.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java [hunk 3]
```diff
@@ -133,6 +133,11 @@ public BaseUpdatePartitionSpec addField(String name, Term term) {
     Pair<Integer, String> validationKey = Pair.of(sourceTransform.first(), sourceTransform.second().toString());
 
     PartitionField existing = transformToField.get(validationKey);
+    if (existing != null && deletes.contains(existing.fieldId()) &&
+        existing.transform().toString().equals(sourceTransform.second().toString())) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:137]: Do we need to compare the `toString` representation? I thought everything implemented `equals` correctly.

**Comment by rdblue** [comment]
Thanks, @CodingCat! Looks close to being ready to commit.

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java (hunk 3 above)
Comment (CodingCat) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java]: updated

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java (hunk 2 above)
Comment (CodingCat) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:145]: updated

File: core/src/test/java/org/apache/iceberg/TestUpdatePartitionSpec.java (hunk 1 above)
Comment (CodingCat) [reply core/src/test/java/org/apache/iceberg/TestUpdatePartitionSpec.java]: it's a good point, updated

File: core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java [hunk 4]
```diff
@@ -124,6 +124,17 @@ public BaseUpdatePartitionSpec addField(Term term) {
...
   }
 
+  private BaseUpdatePartitionSpec rewriteDeleteAndAddField(
+      PartitionField existing, String name, Pair<Integer, Transform<?, ?>> sourceTransform) {
+    deletes.remove(existing.fieldId());
+    if (existing.name().equals(name)) {
+      return this;
+    } else {
+      String newName = name == null ? sourceTransform.second().toString() : name;
+      return renameField(existing.name(), newName);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java:134]: This uses the transform's string representation as the partition field name, but that's not how we generate partition field names. In fact, the result is that we have a partition field name for bucket transforms that is invalid or must be quoted in most systems (`"bucket[16]"`).

Rather than generating the name, I think that this should consider a `null` name a no-op. That is if I remove `shard` and add `bucket(16, id)` then the result should still be named shard -- there's no reason to drop the name and we avoid breaking metadata tables needlessly.

However, if the name was explicitly set, then drop and and should act as a rename.

In short, I think it should be this:

```java
if (name == null || existing.name().equals(name)) {
  return this;
} else {
  return renameField(existing.name(), name);
}
```

Comment (CodingCat) [reply core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java]: makes sense to me, changed accordingly

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Nice work, thanks for fixing this, @CodingCat!
